### PR TITLE
Add top level statuses that can be inferred from PPTS.

### PR DESCRIPTION
**Pull request by rajpara11**
Start creating status history relational table data by creating top-level statuses related to PPTS data. This requires actually looking at the parent-child relationship when it comes to PPTS data. Even when we pull in new PPTS pipeline data we will probably have to still fall back to this logic.

### CODE REVIEW

File: relational/table.py [hunk 1]
```diff
@@ -260,3 +262,154 @@ def rows(self, proj):
...
+        self.gen_id(row, proj)
+        row[self.index(self.TOP_LEVEL_STATUS)] = top_level_status
+        row[self.index(self.START_DATE)] = start_date
+        row[self.index(self.END_DATE)] = end_date
+        row[self.index(self.DATA_SOURCE)] = data
+        return row
+
+    def rows(self, proj):
+        result = []
+        (predev_date, predev_data) = self._predevelopment_date(result, proj)
```
Comment (thelee) [inline relational/table.py]: doesn't look like you're using result in these calls at all - remove?

Comment (rajpara11) [reply relational/table.py]: Done

File: relational/table.py [hunk 2]
```diff
@@ -260,3 +262,154 @@ def rows(self, proj):
...
+
+                date_opened_field = child.get_latest('date_opened')[0]
+                date_opened = datetime.strptime(
+                    date_opened_field.split(' ')[0],
+                    "%m/%d/%Y").date()
+                if date_opened < oldest_open:
+                    oldest_open = date_opened
+
+            if oldest_open < date.max:
+                return (oldest_open.strftime("%Y-%m-%d"), PPTS.OUTPUT_NAME)
```
[the remark this answers is not in the document]
Comment (rajpara11) [reply relational/table.py]: Done

File: relational/table.py [hunk 3]
```diff
@@ -260,3 +262,154 @@ def rows(self, proj):
...
+
+        return ('', None)
+
+    def _filed_for_entitlements_date(self, rows, proj):
+        # TODO: Use the Application Submitted date once we have pulled
+        # in the new PPTS data pipeline (if that doesn't exist fall back to
+        # our own logic)
+
+        # Look for the earliest date_opened on an ENT child of a PRJ.
+        root = proj.roots[PPTS.NAME][0]
```
[the remark this answers is not in the document]
Comment (rajpara11) [reply relational/table.py]: Done

**Comment by rajpara11** [comment]
Yup I've done an end to end run on this (and it's produced a status_history csv with just the statuses I've got so far

**Comment by rajpara11** [comment]
> qq: have you done an end-to-end run on this to generate the new table?



> qq: have you done an end-to-end run on this to generate the new table?

Yup I've done an end to end run on this (and it's produced a status_history csv with just the statuses I've got so far

File: relational/table.py [hunk 4]
```diff
@@ -260,3 +262,159 @@ def rows(self, proj):
...
+
+
+class ProjectStatusHistory(Table):
+    _PPTS_ENT_CODES = {'ENV', 'AHB', 'COA', 'CUA', 'CTZ', 'DNX', 'ENX',
+                       'OFA', 'PTA', 'SHD', 'TDM', 'VAR', 'WLS'}
+
+    TOP_LEVEL_STATUS = 'top_level_status'
+    START_DATE = 'start_date'
+    END_DATE = 'end_date'
+    SEQUENTIAL = 'sequential'
```
Comment (thelee) [inline relational/table.py]: doesn't appear to be used here

Comment (rajpara11) [reply relational/table.py]: Good point and removed, given that we've decided for the time being everythign is sequential probably don't need this

**Pull request merged**